[static runtime] move memonger to new ownership model

bwasti opened the pull request:
Summary:
This works for the contrived example
```
[bwasti@devbig216.ftw3 ~/fbsource/fbcode] buck run mode/opt caffe2/benchmarks/static_runtime:deep_wide_pt_benchmark
Parsing buck files: finished in 0.6 sec
Building: finished in 9.9 sec (100%) 3637/3637 jobs, 0 updated
  Total time: 10.6 sec
Run on (80 X 2001 MHz CPU s)
2020-12-16 09:40:23
--------------------------------------------------------------------------------
Benchmark                                         Time           CPU Iterations
--------------------------------------------------------------------------------
BM_long_static_memory_optimization/2/0         9557 ns       9556 ns      58310
BM_long_static_memory_optimization/8/0         9048 ns       9048 ns      77194
BM_long_static_memory_optimization/32/0       11493 ns      11492 ns      60826
BM_long_static_memory_optimization/512/0    5722748 ns    5722422 ns        122
BM_long_static_memory_optimization/2/1         9007 ns       9006 ns      78032
BM_long_static_memory_optimization/8/1         9012 ns       9012 ns      77415
BM_long_static_memory_optimization/32/1       10639 ns      10638 ns      66973
BM_long_static_memory_optimization/512/1    1279752 ns    1279694 ns        546
```
but I'm not seeing similar speedups on real models yet.

Test Plan:
buck test //caffe2/test:static_runtime
buck test caffe2/benchmarks/static_runtime:static_runtime_cpptest

Differential Revision: D25581156

ezyang commented:
tagging @bdhirsh for use of old operator API

bdhirsh commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -153,8 +154,78 @@ void FuseInferenceOpsForSparseNN(std::shared_ptr<torch::jit::Graph>& graph) {
 #endif
 }
 
+static auto registry =
```
You probably want to update this to the new registration API. It should look something like this:

```
TORCH_LIBRARY(static_runtime, m) {
  m.def("pure_inputs", []() -> at::Tensor { return at::randn({1}); });
  m.def("permute_copy",
    [](at::Tensor self, ArrayRef<int64_t> dims) -> at::Tensor {
                       at::Tensor out = at::empty_like(self);
                       at::native::copy_(out, self);
                       return out.permute(dims);
    });
 }
```

If you have other ops in the `static_runtime` namespace that are registered to the dispatcher, you might hit some problems. The "real" fix would be to migrate all of those ops to use the new registration API and put their definitions all inside of this `TORCH_LIBRARY` block. If that's a big ask though (if the ops are scattered across a lot of bunch of files, etc), then we have a `TORCH_LIBRARY_FRAGMENT` macro that you can use instead (at the cost of some code cleanliness 😄 )

bwasti committed:
[static runtime] Add Alias analysis to Memory Management/Planning (#50060)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/50060

Aliasing is currently mishandled in SR.

This diff fixes that issue entirely and allows us to avoid hard coded "view" registration.  I'll remove the macro in a follow up diff.

However, this diff introduces a subtle assumption when memory optimization is turned on: operators cannot "sometimes alias."  Some care will need to be taken to actually make sure this is enforced going forward.

This diff
```
$ batch=1 ./run.sh
C2 run finished. Milliseconds per iter: 0.0541777. Iters per second: 18457.8
PyTorch run finished. Milliseconds per iter: 0.0669728. Iters per second: 14931.4
$ batch=20 ./run.sh
C2 run finished. Milliseconds per iter: 0.497124. Iters per second: 2011.57
PyTorch run finished. Milliseconds per iter: 0.498913. Iters per second: 2004.36
```

master
```
$ batch=1 ./run.sh
C2 run finished. Milliseconds per iter: 0.0536631. Iters per second: 18634.8
PyTorch run finished. Milliseconds per iter: 0.0672808. Iters per second: 14863.1
$ batch=20 ./run.sh
C2 run finished. Milliseconds per iter: 0.498209. Iters per second: 2007.19
PyTorch run finished. Milliseconds per iter: 0.49974. Iters per second: 2001.04
```

Test Plan:
buck test //caffe2/test:static_runtime
buck test caffe2/benchmarks/static_runtime:static_runtime_cpptest

Differential Revision: D25581156

fbshipit-source-id: ad35bc1777972823d40f7d8f552de353dada5ddb
